OpenVidu/openvidu
# openvidu-browser: rewrite generateOffer() to correctly handle all send/recv cases

1. Uses Transceivers if these are available in the underlying WebRTC implementation; otherwise it falls back to the legacy "offerToReceiveX" config of pc.createOffer().

2. Fixes the multiple audio/video tracks in SDP Offers that originally happened in Safari but in latest master code was happening with all browsers (KMS only supports 1 video and/or 1 audio):

   ```
   WARN  sdpagent  kmssdpagent.c:1557:create_media_answer:<KmsSdpAgent@0x5620b9bf9f60> Cannot handle media 'video UDP/TLS/RTP/SAVPF' (multiple m= lines?)
   ```

3. Fixes the failing video/audio streams on Safari (this was caused by the previous point).

UPDATE 2021-05-19:

* ~Removes the special case for Safari.~ This was already done in master, possibly causing the issue (2).



## Testing procedure

This change should be tested in all [supported platforms](https://docs.openvidu.io/en/2.17.0/troubleshooting/#8-what-platforms-are-supported-by-openvidu), using either the OpenVidu TestApp, OpenVidu Call, or some of the tutorials.


### CI tests
* [x] openvidu-server-e2e (build 5279)
* [x] openvidu-pro-server-e2e (build 2191)



### Test checklist (KMS mode for SDP)

**Desktop browsers**

* [x] Chrome: TestApp + Call, v90.0.4430.212, Linux

**Mobile browsers**

Android:
* [x] Firefox: Call (Juan: OK)
* [ ] Samsung Internet Browser

iOS:
* [ ] Safari (Oscar: FAILS cannot see / be seen by some)

**Mobile native applications**

Android:
* [ ] Ionic
* [ ] Native code SDK

iOS:
* [ ] Ionic
* [ ] Native code SDK

**Desktop applications**

* [ ] Electron



### Test checklist (MEDIASOUP mode for SDP)

**Desktop browsers**

* [x] Chrome: TestApp + Call, v90.0.4430.212, Linux
* [x] Firefox: Call, v88.0, Linux
* [x] Opera: Call, v76.0.4017.123, Windows
* [x] Safari: Call, v14.0, Mac
* [x] Microsoft Edge: Call, v90.0.818.66, Windows

**Mobile browsers**

Android:
* [x] Chrome: Call (Mica: BAD FPS; Juan: OK; Carlos: OK)
* [ ] Firefox: Call (Juan, Pablo: FAILS cannot see / be seen by some)
* [x] Microsoft Edge: Call (Juan: OK)
* [x] Opera: Call (Pablo: OK)
* [x] Samsung Internet Browser: Call (Pablo: OK)

iOS:
* [ ] Safari (Oscar: FAILS cannot see / be seen by some)

**Mobile native applications**

Android:
* [ ] Ionic
* [ ] Native code SDK

iOS:
* [ ] Ionic
* [ ] Native code SDK

**Desktop applications**

* [ ] Electron
